### PR TITLE
Added support for most of <mutex>

### PR DESCRIPTION
**Pull request by ogiroux**
Has mutex, timed_mutex, once_flag, call_once, unique_lock, scoped_lock, and varied free functions that go with them. Excludes only condition variable support and the recursive versions of mutex.

### CODE REVIEW

**Comment by wmaxey** [comment]
TODO: Test ports and heterogeneous tests.

**Comment by jrhemstad** [comment]
@wmaxey is this just waiting on porting tests? I'd like to move forward with exploring https://github.com/NVIDIA/libcudacxx/issues/233, but it depends on having a `mutex` first.

**Comment by miscco** [comment]
@jrhemstad How important is this and should I try to get this rebased?

**Comment by jrhemstad** [comment]
> @jrhemstad How important is this and should I try to get this rebased?

It's important, but not urgent. No need to try and get it merged before the next release. There is still a sizable amount of work to add tests before it can be merged anyways.

File: include/cuda/std/detail/libcxx/include/__mutex_base [hunk 1]
```diff
@@ -37,34 +54,73 @@ _LIBCUDACXX_BEGIN_NAMESPACE_STD
...
+    }
+#endif
+
+    typedef __libcpp_mutex_base_t<_Sco>* native_handle_type;
     _LIBCUDACXX_INLINE_VISIBILITY native_handle_type native_handle() {return &__m_;}
 };
 
-static_assert(is_nothrow_default_constructible<mutex>::value,
-              "the default constructor for std::mutex must be nothrow");
+using mutex = __mutex_base<0>;
```
[the remark this answers is not in the document]
Comment (wmaxey) [reply include/cuda/std/detail/libcxx/include/__mutex_base:121]: Mutex is new, this doesn't change the ABI.

File: .upstream-tests/test/cuda/thread/thread.mutex/thread.mutex.requirements/thread.mutex.requirements.mutex/thread.mutex.class/lock.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+STATIC_TEST_GLOBAL_VAR cuda::mutex<cuda::thread_scope_block>  block_scope;
+STATIC_TEST_GLOBAL_VAR cuda::mutex<cuda::thread_scope_thread> thread_scope;
+
+template<class T> __host__ __device__ T& get_mutex();
+template<> __host__ __device__ cuda::mutex<cuda::thread_scope_system>& get_mutex() { return system_scope; }
+template<> __host__ __device__ cuda::mutex<cuda::thread_scope_device>& get_mutex() { return device_scope; }
+template<> __host__ __device__ cuda::mutex<cuda::thread_scope_block>&  get_mutex() { return block_scope; }
+template<> __host__ __device__ cuda::mutex<cuda::thread_scope_thread>& get_mutex() { return thread_scope; }
+
+using Clock = cuda::std::chrono::system_clock;
```
[the remark this answers is not in the document]
Comment (gonzalobg) [reply .upstream-tests/test/cuda/thread/thread.mutex/thread.mutex.requirements/thread.mutex.requirements.mutex/thread.mutex.class/lock.pass.cpp:39]: This was resolved, but I still see the typedefs below.

File: include/cuda/std/detail/libcxx/include/__mutex_base [hunk 3]
```diff
@@ -274,6 +345,8 @@ void
 swap(unique_lock<_Mutex>& __x, unique_lock<_Mutex>& __y) _NOEXCEPT
     {__x.swap(__y);}
 
+#ifndef _LIBCUDACXX_HAS_THREAD_API_CUDA
```
Comment (gonzalobg) [inline include/cuda/std/detail/libcxx/include/__mutex_base:348]: Is condition variable supported?

**Comment by miscco** [comment]
Dropping in favor of https://github.com/NVIDIA/cccl/pull/187

**Pull request closed without merging**